File: solutions/2026-08-02/javascript/2026-08-02-stone-game.js
/**
 * @param {number[]} piles
 * @return {boolean}
 */

/*
Problem Summary: Alice and Bob play a game with an even number of stone piles. They take turns taking a pile from either end. The player with more stones wins. Alice goes first.
Link: https://leetcode.com/problems/stone-game/

Approach:
This is a classic game theory problem that can be solved using dynamic programming. Since the total number of stones is odd and there are an even number of piles, there will always be a winner, and no ties. Alice goes first.

The core idea is to determine the maximum difference in scores Alice can achieve over Bob. If Alice can guarantee a positive difference, she wins.

We can use a 2D DP array `dp[i][j]` to store the maximum difference in scores a player can achieve when considering the piles from index `i` to `j` (inclusive).

The state transition would be:
When it's a player's turn to pick from `piles[i...j]`:
1. If the player picks `piles[i]`, their score increases by `piles[i]`, and the remaining piles are `piles[i+1...j]`. The opponent will then play optimally on these remaining piles, aiming to maximize *their* score difference. From the current player's perspective, this means their score difference will be `piles[i] - dp[i+1][j]`.
2. If the player picks `piles[j]`, their score increases by `piles[j]`, and the remaining piles are `piles[i...j-1]`. The opponent will then play optimally on these remaining piles. From the current player's perspective, this means their score difference will be `piles[j] - dp[i][j-1]`.

The player will choose the move that maximizes their score difference:
`dp[i][j] = max(piles[i] - dp[i+1][j], piles[j] - dp[i][j-1])`

Base cases:
- When `i == j` (a single pile), the player takes that pile, and the difference is `piles[i]`. So, `dp[i][i] = piles[i]`.

The DP table will be filled diagonally, starting with length 1 subarrays, then length 2, and so on, up to the full array.

Since Alice starts, we are interested in `dp[0][n-1]`, which represents the maximum score difference Alice can achieve over Bob for the entire array of piles. If `dp[0][n-1] > 0`, Alice wins.

However, a key observation for this specific problem is that because the number of piles is even and Alice goes first, Alice can always choose to take either all the odd-indexed piles or all the even-indexed piles.
Let's say Alice decides to take all the odd-indexed piles (0-indexed: 0, 2, 4, ...). Bob will be left with the even-indexed piles (1, 3, 5, ...).
Similarly, Alice can decide to take all the even-indexed piles (1, 3, 5, ...). Bob will be left with the odd-indexed piles (0, 2, 4, ...).

Since Alice can choose her strategy beforehand and play optimally, she can always ensure she gets more stones than Bob. The total sum of stones is odd, so there are no ties. Therefore, Alice will always win.

This means the answer is always true. The problem is designed to test if you recognize this game theory aspect rather than just implementing DP.
If we were to implement the DP, it would look like this:

Let n be the number of piles.
Initialize dp[n][n] with 0.

For length = 1 to n:
  For i = 0 to n - length:
    j = i + length - 1
    If length == 1:
      dp[i][j] = piles[i]
    Else:
      dp[i][j] = max(piles[i] - dp[i+1][j], piles[j] - dp[i][j-1])

Return dp[0][n-1] > 0.

Given the problem constraints and the nature of the game, Alice *always* wins. This is because she can choose to take either all the piles at even indices or all the piles at odd indices. She can calculate the total sum of stones at even indices and the total sum of stones at odd indices and choose the strategy that yields more stones. Since the total sum of stones is odd, one of these sums must be greater than the other.

Time Complexity: O(1) - Because the problem guarantees Alice wins due to the game's structure. If we were to implement the DP solution: O(n^2), where n is the number of piles.
Space Complexity: O(1) - Because the problem guarantees Alice wins. If we were to implement the DP solution: O(n^2) for the DP table.

Given the context of competitive programming problems, sometimes there's a trick or a mathematical property that simplifies the solution significantly. This problem is one such case. Alice can always win.
*/

var stoneGame = function(piles) {
    // In this game, Alice always wins.
    // Here's why:
    // The total number of piles is even (let's say 2k).
    // Alice goes first.
    // Alice can choose to take only piles at even indices (0, 2, 4, ...)
    // OR Alice can choose to take only piles at odd indices (1, 3, 5, ...).
    //
    // Consider two scenarios:
    // 1. Alice aims for even-indexed piles, Bob gets odd-indexed piles.
    // 2. Alice aims for odd-indexed piles, Bob gets even-indexed piles.
    //
    // Alice can calculate the sum of stones at all even indices and the sum of stones
    // at all odd indices.
    //
    // Let S_even = sum of piles[i] where i is even.
    // Let S_odd = sum of piles[i] where i is odd.
    //
    // Total stones = S_even + S_odd.
    // The problem states the total sum is odd, so S_even != S_odd.
    //
    // Alice can choose to target the set of piles (either even-indexed or odd-indexed)
    // that has a larger sum.
    //
    // For example, if S_even > S_odd:
    // Alice strategy:
    // On her first turn, if piles[0] is part of the "even" set she wants and
    // piles[n-1] is part of the "odd" set Bob would get, she takes piles[0].
    // This leaves piles[1...n-1]. Bob now has to pick from piles[1] or piles[n-1].
    //
    // The key is that Alice, by making her initial choice (take from left or right),
    // dictates which "parity" of pile index she will eventually have access to on
    // subsequent turns.
    //
    // More formally:
    // Alice can always ensure she gets either all the piles originally at even indices
    // or all the piles originally at odd indices.
    //
    // Example: piles = [5, 3, 4, 5]
    // Even indices: piles[0]=5, piles[2]=4. Total = 9.
    // Odd indices: piles[1]=3, piles[3]=5. Total = 8.
    //
    // Alice can choose to target the even-indexed piles (sum 9).
    // Turn 1 (Alice): She wants the 5 at index 0 or the 4 at index 2.
    // If she takes piles[0] (5), remaining: [3, 4, 5]. Bob's turn.
    //   Bob can take 3 or 5.
    //   If Bob takes 3, remaining: [4, 5]. Alice wants the 4 (even index). Alice takes 4. Alice's total: 5+4=9. Bob's total: 3+5=8. Alice wins.
    //   If Bob takes 5, remaining: [3, 4]. Alice wants the 4 (even index). Alice takes 4. Alice's total: 5+4=9. Bob's total: 5+3=8. Alice wins.
    //
    // If Alice takes piles[3] (5) initially, remaining: [5, 3, 4]. Bob's turn.
    //   Bob can take 5 or 4.
    //   If Bob takes 5, remaining: [3, 4]. Alice wants the 4. Alice takes 4. Alice's total: 5+4=9. Bob's total: 5+3=8. Alice wins.
    //   If Bob takes 4, remaining: [5, 3]. Alice wants the 5. Alice takes 5. Alice's total: 5+5=10. Bob's total: 4+3=7. Alice wins.
    //
    // In all optimal play scenarios, Alice can secure more stones.
    // Therefore, Alice always wins.

    return true;

    /*
    // This is the DP approach if one were to solve it without the game theory insight.
    // It's more complex than needed for this specific LeetCode problem.

    const n = piles.length;
    // dp[i][j] will store the maximum difference in score the current player
    // can achieve over the other player when considering piles from index i to j.
    // The current player is the one whose turn it is to pick from piles[i...j].
    const dp = Array(n).fill(0).map(() => Array(n).fill(0));

    // Base case: When there's only one pile, the current player takes it.
    for (let i = 0; i < n; i++) {
        dp[i][i] = piles[i];
    }

    // Fill the dp table diagonally.
    // 'len' represents the length of the subarray of piles being considered.
    for (let len = 2; len <= n; len++) {
        // 'i' is the starting index of the subarray.
        for (let i = 0; i <= n - len; i++) {
            // 'j' is the ending index of the subarray.
            const j = i + len - 1;

            // Option 1: The current player takes the pile at index 'i'.
            // Their score increases by piles[i].
            // The opponent then plays on the remaining piles from i+1 to j.
            // The opponent will try to maximize their score, meaning they will
            // achieve a difference of dp[i+1][j] from their perspective.
            // From the current player's perspective, this difference is piles[i] - dp[i+1][j].
            const takeLeft = piles[i] - dp[i + 1][j];

            // Option 2: The current player takes the pile at index 'j'.
            // Their score increases by piles[j].
            // The opponent then plays on the remaining piles from i to j-1.
            // From the current player's perspective, this difference is piles[j] - dp[i][j-1].
            const takeRight = piles[j] - dp[i][j - 1];

            // The current player chooses the option that maximizes their score difference.
            dp[i][j] = Math.max(takeLeft, takeRight);
        }
    }

    // dp[0][n-1] stores the maximum score difference Alice can achieve over Bob
    // for the entire array of piles. If this difference is positive, Alice wins.
    return dp[0][n - 1] > 0;
    */
};
```

File: solutions/2026-08-02/java/2026-08-02-stone-game.java
// Summary: Alice and Bob play a game with an even number of stone piles. They take turns choosing a pile from either end. Alice wins if she has more stones than Bob, assuming optimal play.
// Link: https://leetcode.com/problems/stone-game/
// Approach: This is a classic game theory problem that can be solved using dynamic programming.
// The core idea is to define a state `dp[i][j]` which represents the maximum difference in score that the current player can achieve
// when considering the piles from index `i` to `j` (inclusive).
//
// For each subproblem `dp[i][j]`, the current player has two choices:
// 1. Take the pile at index `i`: The player gets `piles[i]` stones. The remaining piles are from `i+1` to `j`. The next player will play on this subarray.
//    The score difference for the current player will be `piles[i] - dp[i+1][j]`.
// 2. Take the pile at index `j`: The player gets `piles[j]` stones. The remaining piles are from `i` to `j-1`. The next player will play on this subarray.
//    The score difference for the current player will be `piles[j] - dp[i][j-1]`.
//
// The current player will choose the option that maximizes their score difference:
// `dp[i][j] = max(piles[i] - dp[i+1][j], piles[j] - dp[i][j-1])`
//
// The base cases are when `i == j`. In this case, `dp[i][i] = piles[i]`.
// We iterate through increasing lengths of subarrays (from 1 to `n`).
// The final answer will be `dp[0][n-1]`. If `dp[0][n-1] > 0`, it means Alice (the first player) can achieve a positive score difference, and thus wins.
//
// IMPORTANT OBSERVATION:
// Due to the problem constraints (even number of piles, total odd number of stones), Alice always has a winning strategy.
// Alice can choose to always pick from either the even-indexed piles or the odd-indexed piles.
// Let's consider the sum of stones at even indices and odd indices.
// Sum_even = piles[0] + piles[2] + ... + piles[n-2]
// Sum_odd  = piles[1] + piles[3] + ... + piles[n-1]
// Total sum is Sum_even + Sum_odd, which is odd. This means Sum_even != Sum_odd.
//
// Alice can decide at the beginning whether she wants to target the set of piles with a larger sum.
// For example, if Sum_even > Sum_odd, Alice can ensure she always picks from an even-indexed pile.
// If Alice picks piles[0], the remaining piles are from index 1 to n-1. The next player (Bob) faces piles at indices 1, 2, ..., n-1.
// If Bob picks piles[1] (odd index), Alice can pick piles[2] (even index).
// If Bob picks piles[n-1] (odd index), Alice can pick piles[n-2] (even index).
// Alice can always maintain the strategy of picking from the 'preferred' parity of indices.
// Since Sum_even != Sum_odd, Alice can always pick the parity that gives her more stones.
// Therefore, Alice always wins.
//
// Time Complexity: O(n^2) for the DP approach. However, due to the mathematical property explained above, the actual time complexity to determine the winner is O(1).
// Space Complexity: O(n^2) for the DP approach. However, due to the mathematical property explained above, the actual space complexity to determine the winner is O(1).

class Solution {
    public boolean stoneGame(int[] piles) {
        // Due to the problem constraints (even number of piles, total odd number of stones),
        // Alice always has a winning strategy by picking from piles of a specific parity
        // (either all even-indexed piles or all odd-indexed piles).
        // Since the total number of stones is odd, the sum of stones at even indices
        // and the sum of stones at odd indices cannot be equal.
        // Alice can choose to target the set of piles with the larger sum, and she can
        // always ensure she picks from this set.
        // Therefore, Alice always wins.
        return true;

        /*
        // DP approach for completeness (though not necessary for this problem given the constraints)
        int n = piles.length;
        // dp[i][j] will store the maximum score difference the current player can achieve
        // from piles[i] to piles[j]
        int[][] dp = new int[n][n];

        // Base case: when only one pile is left, the player takes it.
        for (int i = 0; i < n; i++) {
            dp[i][i] = piles[i];
        }

        // Fill the DP table for lengths of subarrays from 2 to n
        for (int len = 2; len <= n; len++) {
            for (int i = 0; i <= n - len; i++) {
                int j = i + len - 1;
                // The current player can either take piles[i] or piles[j].
                // If they take piles[i], their score is piles[i] plus the remaining score difference from dp[i+1][j].
                // The difference is positive if they take more than the opponent.
                // The score difference achieved by the *current* player from the remaining subarray is dp[i+1][j].
                // So, if the current player takes piles[i], their score gain relative to the opponent is piles[i] - dp[i+1][j].
                // Similarly, if they take piles[j], their score gain is piles[j] - dp[i][j-1].
                // The current player will choose the move that maximizes this difference.
                dp[i][j] = Math.max(piles[i] - dp[i + 1][j], piles[j] - dp[i][j - 1]);
            }
        }

        // If dp[0][n-1] is positive, it means Alice (the first player) can achieve a positive score difference,
        // thus she wins.
        return dp[0][n - 1] > 0;
        */
    }
}
